File: lisca-rs/src/common/progress.rs
use std::cell::RefCell;
use std::io::{self, IsTerminal, Write};
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};

use serde::{Deserialize, Serialize};

const DEFAULT_BAR_WIDTH: usize = 28;
const DEFAULT_PLAIN_INTERVAL: Duration = Duration::from_millis(250);

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ProgressKind {
    Start,
    #[default]
    Update,
    Finish,
    Error,
}

#[derive(Clone, Debug, Default, Serialize, Deserialize)]
pub struct ProgressEvent {
    #[serde(default)]
    pub kind: ProgressKind,
    pub progress: f64,
    pub message: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub step: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub current: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub total: Option<u64>,
    #[serde(default)]
    pub timestamp: String,
}

#[derive(Debug)]
struct RendererState {
    last_line_len: usize,
    last_plain_progress: f64,
    last_plain_message: String,
    last_plain_at: Option<Instant>,
}

impl Default for RendererState {
    fn default() -> Self {
        Self {
            last_line_len: 0,
            last_plain_progress: -1.0,
            last_plain_message: String::new(),
            last_plain_at: None,
        }
    }
}

pub struct TerminalProgressRenderer {
    interactive: bool,
    bar_width: usize,
    plain_interval: Duration,
    state: RefCell<RendererState>,
}

impl TerminalProgressRenderer {
    pub fn stderr() -> Self {
        Self::new(io::stderr().is_terminal())
    }

    pub fn new(interactive: bool) -> Self {
        Self {
            interactive,
            bar_width: DEFAULT_BAR_WIDTH,
            plain_interval: DEFAULT_PLAIN_INTERVAL,
            state: RefCell::new(RendererState::default()),
        }
    }

    pub fn handle(&self, event: ProgressEvent) {
        let event = normalize(event);
        if self.interactive {
            self.render_interactive(&event);
        } else {
            self.render_plain(&event);
        }
    }

    fn render_interactive(&self, event: &ProgressEvent) {
        let progress = clamp_progress(event.progress);
        let filled = ((progress * self.bar_width as f64).round() as usize).min(self.bar_width);
        let bar = format!(
            "{}{}",
            "#".repeat(filled),
            "-".repeat(self.bar_width.saturating_sub(filled))
        );
        let percent = (progress * 100.0).round() as u32;
        let line = format!(
            "{percent:>3}% [{bar}] {}{}",
            event.message,
            event_suffix(event)
        );

        let mut state = self.state.borrow_mut();
        let padding = state.last_line_len.saturating_sub(line.len());
        let mut stderr = io::stderr();
        let _ = write!(stderr, "\r{line}{}", " ".repeat(padding));
        let _ = stderr.flush();

        let terminal =
            matches!(event.kind, ProgressKind::Finish | ProgressKind::Error) || progress >= 1.0;
        state.last_line_len = if terminal { 0 } else { line.len() };
        if terminal {
            let _ = writeln!(stderr);
            let _ = stderr.flush();
        }
    }

    fn render_plain(&self, event: &ProgressEvent) {
        let progress = clamp_progress(event.progress);
        let percent = (progress * 100.0).round() as u32;
        let now = Instant::now();
        let mut state = self.state.borrow_mut();
        let should_print = matches!(
            event.kind,
            ProgressKind::Start | ProgressKind::Finish | ProgressKind::Error
        ) || event.message != state.last_plain_message
            || (progress - state.last_plain_progress).abs() >= 0.05
            || state
                .last_plain_at
                .map(|last| now.duration_since(last) >= self.plain_interval)
                .unwrap_or(true);
        if !should_print {
            return;
        }

        let mut stderr = io::stderr();
        let _ = writeln!(
            stderr,
            "{:>6} {percent:>3}% {}{}",
            kind_label(event.kind),
            event.message,
            event_suffix(event)
        );
        let _ = stderr.flush();

        state.last_plain_progress = progress;
        state.last_plain_message = event.message.clone();
        state.last_plain_at = Some(now);
    }
}

pub fn progress_event(progress: f64, message: impl Into<String>) -> ProgressEvent {
    ProgressEvent {
        kind: ProgressKind::Update,
        progress: clamp_progress(progress),
        message: message.into(),
        step: None,
        current: None,
        total: None,
        timestamp: progress_timestamp(),
    }
}

pub fn start(progress: f64, message: impl Into<String>) -> ProgressEvent {
    ProgressEvent {
        kind: ProgressKind::Start,
        ..progress_event(progress, message)
    }
}

pub fn finish(progress: f64, message: impl Into<String>) -> ProgressEvent {
    ProgressEvent {
        kind: ProgressKind::Finish,
        ..progress_event(progress, message)
    }
}

pub fn error(progress: f64, message: impl Into<String>) -> ProgressEvent {
    ProgressEvent {
        kind: ProgressKind::Error,
        ..progress_event(progress, message)
    }
}

pub fn legacy_adapter<'a>(sink: &'a dyn Fn(ProgressEvent)) -> impl Fn(f64, &str) + 'a {
    move |progress, message| sink(progress_event(progress, message))
}

pub fn progress_timestamp() -> String {
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default();
    format!("{}.{:09}Z", now.as_secs(), now.subsec_nanos())
}

fn clamp_progress(progress: f64) -> f64 {
    progress.clamp(0.0, 1.0)
}

fn normalize(mut event: ProgressEvent) -> ProgressEvent {
    event.progress = clamp_progress(event.progress);
    if event.timestamp.is_empty() {
        event.timestamp = progress_timestamp();
    }
    event
}

fn event_suffix(event: &ProgressEvent) -> String {
    if let (Some(current), Some(total)) = (event.current, event.total) {
        return format!(" ({current}/{total})");
    }
    if let Some(step) = event.step.as_deref() {
        return format!(" [{step}]");
    }
    String::new()
}

fn kind_label(kind: ProgressKind) -> &'static str {
    match kind {
        ProgressKind::Start => "start",
        ProgressKind::Update => "update",
        ProgressKind::Finish => "finish",
        ProgressKind::Error => "error",
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::cell::RefCell;

    #[test]
    fn legacy_adapter_emits_update_events() {
        let events = RefCell::new(Vec::new());
        {
            let sink = |event: ProgressEvent| events.borrow_mut().push(event);
            let callback = legacy_adapter(&sink);
            callback(0.4, "Processing");
        }

        let events = events.into_inner();
        assert_eq!(events.len(), 1);
        assert_eq!(events[0].kind, ProgressKind::Update);
        assert_eq!(events[0].progress, 0.4);
        assert_eq!(events[0].message, "Processing");
    }

    #[test]
    fn finish_event_uses_finish_kind() {
        let event = finish(1.0, "Done");
        assert_eq!(event.kind, ProgressKind::Finish);
        assert_eq!(event.progress, 1.0);
        assert_eq!(event.message, "Done");
    }
}

File: lisca-py/src/lisca_py/common/progress.py
from __future__ import annotations

import sys
import time
from collections.abc import Callable
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import StrEnum
from typing import Protocol, TextIO

ProgressCallback = Callable[[float, str], None]


class ProgressKind(StrEnum):
    START = "start"
    UPDATE = "update"
    FINISH = "finish"
    ERROR = "error"


@dataclass(slots=True)
class ProgressEvent:
    kind: ProgressKind
    progress: float
    message: str
    step: str | None = None
    current: int | None = None
    total: int | None = None
    timestamp: str = field(default_factory=lambda: progress_timestamp())


class ProgressSink(Protocol):
    def __call__(self, event: ProgressEvent, /) -> None: ...


def progress_timestamp() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def clamp_progress(progress: float) -> float:
    return max(0.0, min(1.0, progress))


def progress_event(
    progress: float,
    message: str,
    *,
    kind: ProgressKind = ProgressKind.UPDATE,
    step: str | None = None,
    current: int | None = None,
    total: int | None = None,
    timestamp: str | None = None,
) -> ProgressEvent:
    return ProgressEvent(
        kind=kind,
        progress=clamp_progress(progress),
        message=message,
        step=step,
        current=current,
        total=total,
        timestamp=timestamp or progress_timestamp(),
    )


def legacy_callback_adapter(sink: ProgressSink | None) -> ProgressCallback | None:
    if sink is None:
        return None

    def callback(progress: float, message: str) -> None:
        sink(progress_event(progress, message))

    return callback


class TerminalProgressRenderer:
    def __init__(
        self,
        stream: TextIO | None = None,
        *,
        interactive: bool | None = None,
        plain_interval: float = 0.25,
        bar_width: int = 28,
    ) -> None:
        self._stream = stream or sys.stderr
        isatty = getattr(self._stream, "isatty", None)
        self._interactive = interactive if interactive is not None else bool(callable(isatty) and isatty())
        self._plain_interval = plain_interval
        self._bar_width = bar_width
        self._last_line_length = 0
        self._last_plain_progress = -1.0
        self._last_plain_message = ""
        self._last_plain_at = 0.0

    def __call__(self, event: ProgressEvent) -> None:
        normalized = progress_event(
            event.progress,
            event.message,
            kind=event.kind,
            step=event.step,
            current=event.current,
            total=event.total,
            timestamp=event.timestamp,
        )
        if self._interactive:
            self._render_interactive(normalized)
            return
        self._render_plain(normalized)

    def _render_interactive(self, event: ProgressEvent) -> None:
        progress = clamp_progress(event.progress)
        filled = min(self._bar_width, int(round(progress * self._bar_width)))
        bar = "#" * filled + "-" * (self._bar_width - filled)
        percent = int(round(progress * 100))
        suffix = self._event_suffix(event)
        line = f"{percent:>3}% [{bar}] {event.message}{suffix}"
        padding = max(0, self._last_line_length - len(line))
        self._stream.write(f"\r{line}{' ' * padding}")
        self._stream.flush()
        terminal = event.kind in {ProgressKind.FINISH, ProgressKind.ERROR} or progress >= 1.0
        self._last_line_length = 0 if terminal else len(line)
        if terminal:
            self._stream.write("\n")
            self._stream.flush()

    def _render_plain(self, event: ProgressEvent) -> None:
        progress = clamp_progress(event.progress)
        now = time.monotonic()
        should_print = (
            event.kind in {ProgressKind.START, ProgressKind.FINISH, ProgressKind.ERROR}
            or event.message != self._last_plain_message
            or abs(progress - self._last_plain_progress) >= 0.05
            or now - self._last_plain_at >= self._plain_interval
        )
        if not should_print:
            return
        percent = int(round(progress * 100))
        suffix = self._event_suffix(event)
        self._stream.write(f"{event.kind.value:>6} {percent:>3}% {event.message}{suffix}\n")
        self._stream.flush()
        self._last_plain_message = event.message
        self._last_plain_progress = progress
        self._last_plain_at = now

    @staticmethod
    def _event_suffix(event: ProgressEvent) -> str:
        if event.current is not None and event.total is not None:
            return f" ({event.current}/{event.total})"
        if event.step:
            return f" [{event.step}]"
        return ""


def progress_terminal_stderr() -> ProgressSink:
    return TerminalProgressRenderer(sys.stderr)
